assets/src/edit-story/components/canvas/useInsertTextSet.js

```diff
@@ -60,40 +62,32 @@ function useInsertTextSet() {
   );
 
   const insertTextSetByOffset = useCallback(
-    (elements, { offsetX, offsetY }, boundary) => {
+    (elements, { offsetX, offsetY }) => {
       if (!elements.length) {
         return;
       }
 
-      let adjustedOffsetX = offsetX;
-      let adjustedOffsetY = offsetY;
-
-      if (boundary) {
-        const { width, height } = boundary;
-        const { textSetWidth, textSetHeight } = elements[0];
-
-        if (offsetX < 0) {
-          adjustedOffsetX = 0;
-        } else if (offsetX > width) {
-          adjustedOffsetX = width - textSetWidth;
-        } else if (offsetX + textSetWidth > width) {
-          adjustedOffsetX -= offsetX + textSetWidth - width;
-        }
-
-        if (offsetY < 0) {
-          adjustedOffsetY = 0;
-        } else if (offsetY > height) {
-          adjustedOffsetY = height - textSetHeight;
-        } else if (offsetY + textSetHeight > height) {
-          adjustedOffsetY -= offsetY + textSetHeight - height;
-        }
-      }
-
-      const positionedTextSet = elements.map((element) => ({
-        ...element,
-        x: element.normalizedOffsetX + adjustedOffsetX,
-        y: element.normalizedOffsetY + adjustedOffsetY,
-      }));
+      const positionedTextSet = elements
+        .map((element) => {
+          // Skip adding any elements that are outside of page.
+          const x = element.normalizedOffsetX + offsetX;
+          const y = element.normalizedOffsetY + offsetY;
+          const { width, height } = element;
+          if (
+            x > PAGE_WIDTH ||
+            x + width < 0 ||
+            y > FULLBLEED_HEIGHT ||
+            y + height < -DANGER_ZONE_HEIGHT
+          ) {
+            return null;
+          }
+          return {
+            ...element,
+            x,
+            y,
+          };
+        })
+        .filter((el) => el);
```

Comment:
I think this is a super clean way to do this, is it expected for items dropped outside of the safe zone to be eliminated? If so, then great!

Reply:
Items outside of the Page are expected to be removed -- it's ok to drop outside of the safe zone as long as it's partially on the Page but nothing outside of the Page should get added. That's the general expected logic (as of now) when adding elements / moving elements, might need a follow-up issue to discuss with UX if the expected behavior for text sets seems to make sense in a different way.